### PR TITLE
feat(pairing): The channel server for the pairing flow.

### LINKED ISSUE #6613: Land the channel-server client for the pairing flow

### PR DESCRIPTION
**Pull request by shane-tomlinson**
fixes #6613

### CODE REVIEW

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 0784ca2076c9c5c1003dc9fc3e7270452a14db21
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/mozilla/fxa-content-server/blob/master/CONTRIBUTING.md#git-commit-guidelines

---
This message was auto-generated by https://gitcop.com

**Comment by shane-tomlinson** [comment]
@mozilla/fxa-devs - r?

File: app/scripts/lib/channel-server-client.js [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+   * Use `channelKeyBuffer` and `channelIdBuffer` to derive
+   * the encryption/decryption JWK for the channel
+   *
+   * @param {Buffer} channelKeyBuffer
+   * @param {Buffer} channelIdBuffer
+   * @returns {Promise} resolves to the JWK when complete
+   * @private
+   */
+  _deriveChannelJWK(channelKeyBuffer, channelIdBuffer) {
+    return hkdf(channelKeyBuffer, channelIdBuffer, CHANNEL_KEY_INFO_BUFFER, DEVICE_PAIRING_CHANNEL_KEY_BYTES)
```
Comment (shane-tomlinson) [inline app/scripts/lib/channel-server-client.js]: @rfk, @eoger - is this how you'd expect the channel JWK and confirmation code to be derived?

File: app/scripts/lib/channel-server-client.js [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+
+import a256gcm from './crypto/a256gcm';
+import base64url from 'base64url';
+import ChannelServerClientErrors from './channel-server-client-errors';
+import { DEVICE_PAIRING_CHANNEL_KEY_BYTES } from './constants';
+import { Model } from 'backbone';
+import { pick } from 'underscore';
+import hkdf from './crypto/hkdf';
+
+const CHANNEL_KEY_INFO_BUFFER = Buffer.from('identity.mozilla.com/picl/v1/pair/encryption-key', 'utf8');
```
Comment (vladikoff) [inline app/scripts/lib/channel-server-client.js]: What provides `Buffer` for us?

File: app/scripts/lib/channel-server-client.js [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+   * {
+   *   "message": <message name>,
+   *   "data": <any, specific to the message>
+   * }
+   *
+   * @param {Object} event
+   * @private
+   */
+  _encryptedMessageHandler (event) {
+    const { message: ciphertext, sender } = JSON.parse(event.data);
```
Comment (vladikoff) [inline app/scripts/lib/channel-server-client.js]: do we want to handle JSON.parse fails here? or just let it throw a generic parse error?

File: app/scripts/lib/channel-server-client.js (hunk 1 above)
Comment (eoger) [reply app/scripts/lib/channel-server-client.js]: LGTM

File: app/scripts/lib/channel-server-client.js [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+
+  /**
+   * Get the encryption/decryption JWK for the channel
+   *
+   * @returns {Promise} resolves to the JWK when complete
+   * @private
+   */
+  _getChannelJWK () {
+    const { channelId, channelJWK, channelKey } = this.toJSON();
+    if (channelJWK) {
```
Comment (vladikoff) [inline app/scripts/lib/channel-server-client.js]: different styles for `JWK` vs `Jwk`, 

see `channelJWK` and `keysJwk` above.

File: app/tests/spec/lib/channel-server-client.js [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+
+    it('hooks up the expected handlers', () => {
+      const expectedEvents = [
+        'close',
+        'error',
+        'message',
+        'open',
+      ];
+
+      //assert.equal(socket.addEventListener.callCount, expectedEvents.length);
```
Comment (vladikoff) [inline app/tests/spec/lib/channel-server-client.js]: 👁

File: app/scripts/lib/channel-server-client.js (hunk 3 above)
Comment (shane-tomlinson) [reply app/scripts/lib/channel-server-client.js]: We probably should throw a better error to easily find them in Sentry

File: app/scripts/lib/channel-server-client.js (hunk 2 above)
Comment (shane-tomlinson) [reply app/scripts/lib/channel-server-client.js]: > What provides Buffer for us?

I'm going to assume it's the [Buffer implementation in node-libs-browser](https://github.com/webpack/node-libs-browser), is that the same as the one used by the crypto-relier?

File: app/tests/spec/lib/channel-server-client.js (hunk 5 above)
Comment (shane-tomlinson) [reply app/tests/spec/lib/channel-server-client.js]: Good catch.

File: app/scripts/lib/channel-server-client.js [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+   * @param {Object} event
+   * @returns {Promise} resolves when complete, rejects if unable to parse or
+   *   decrypt the message.
+   * @private
+   */
+  _encryptedMessageHandler (event) {
+    return this._parseMessageEvent(event)
+      .then(({ data, message }) => {
+        this.trigger(message, data);
+        this._notifier.trigger(message, data);
```
[the remark this answers is not in the document]
Comment (shane-tomlinson) [reply app/scripts/lib/channel-server-client.js]: I don't think the notifier is actually used by the rest of the pairing code. I'm going to remove it as well as add a `pairing:remote:` namespace to all of the messages. I'm also going to ensure there *is* a message.

File: app/tests/spec/lib/channel-server-client.js [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+
+  describe('_deriveChannelJwk', () => {
+    it('creates a Jwk', () => {
+      const channelKeyBuffer = Buffer.from('channel id', 'utf8');
+      const channelIdBuffer = Buffer.from('channel key', 'utf8');
+      return client._deriveChannelJwk(channelKeyBuffer, channelIdBuffer)
+        .then(jwk => {
+          assert.equal(jwk.alg, 'A256GCM');
+          assert.equal(jwk.kty, 'oct');
+          assert.ok(jwk.kid);
```
[the remark this answers is not in the document]
Comment (shane-tomlinson) [reply app/tests/spec/lib/channel-server-client.js]: I will add these tests, however, for the confirmation code, we spec'd the info string to be `identity.mozilla.com/picl/v1/pair/confirmation-code`, with a result of `d45a83f6` in both Python and JS.

[the remark this answers is not in the document]
Comment (shane-tomlinson) [reply app/tests/spec/lib/channel-server-client.js]: > HKDF("channel key", "channel id", "identity.mozilla.com/picl/v1/pair/encryption-key", 32).encode('hex')
'0d91a09d8e04f34cdd3045026eb98d96c3e999908f58956fdba6a39b685472f8'

We can use this to ensure we derive the correct channel key. However, IIUC, the value of [jwk.kid is implementation specific](https://tools.ietf.org/html/rfc7517#section-4.5).

File: app/tests/spec/lib/channel-server-client.js [hunk 8]
```diff
@@ -5,55 +5,93 @@
...
+        const jwkObj = jwk.toObject(true);
+        assert.equal(jwkObj.alg, 'A256GCM');
+        assert.equal(jwkObj.kty, 'oct');
+        assert.equal(jwkObj.k.toString('hex'), expectedKeyBuffer.toString('hex'));
+        assert.ok(jwk.kid);
+      });
+    });
+  });
+
+  describe('_deriveEncryptionKey', () => {
```
Comment (rfk) [inline app/tests/spec/lib/channel-server-client.js]: This appears to be identical to the `_deriveEncryptionKey' test above, is it an accidental dup?

Comment (shane-tomlinson) [reply app/tests/spec/lib/channel-server-client.js]: It is.

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 7f8c779eeda03dfa2deee2056b0e7fc2630f9240
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 2515e8b0633eb35b8945bf9999b92cafb996ae23
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/mozilla/fxa-content-server/blob/master/CONTRIBUTING.md#git-commit-guidelines

---
This message was auto-generated by https://gitcop.com

**Pull request merged**